File: parsers/csv_parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
CSV file format parser
"""

import csv
import sys
from typing import List, Dict, Any
from .base import BaseParser


class CSVParser(BaseParser):
    """Parser for CSV files with header row"""

    format_name = "csv"
    file_extensions = ['.csv', '.tsv']
    mime_types = ['text/csv', 'text/tab-separated-values']

    def supports_format(self, file_path: str) -> bool:
        """Check if file is CSV"""
        ext = file_path.lower().split('.')[-1]
        return ext in ['csv', 'tsv']

    def load(self, file_path: str) -> List[Dict[str, Any]]:
        """Load CSV file and convert to list of dictionaries"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                # Auto-detect delimiter
                sample = f.read(1024)
                f.seek(0)

                delimiter = ','
                try:
                    sniffer = csv.Sniffer()
                    if sniffer.has_header(sample):
                        delimiter = sniffer.sniff(sample).delimiter
                except csv.Error:
                    delimiter = ','

                # Read with detected delimiter
                reader = csv.DictReader(f, delimiter=delimiter)
                data = list(reader)

                # Type inference: try to convert strings to numbers
                for row in data:
                    for key, value in row.items():
                        row[key] = self._infer_type(value)

                return data

        except FileNotFoundError:
            print(f"错误: 文件 {file_path} 不存在")
            sys.exit(1)
        except Exception as e:
            print(f"读取CSV错误: {e}")
            sys.exit(1)

    def _infer_type(self, value: str) -> Any:
        """Try to convert string to int, float, or keep as string"""
        try:
            if '.' in value:
                return float(value)
            return int(value)
        except (ValueError, TypeError):
            return value

File: nodejs/src/parsers/csv_parser.ts
import * as fs from 'fs';
import { BaseParser, ParserData } from './base';

export class CSVParser extends BaseParser {
  formatName = 'csv';
  fileExtensions = ['.csv', '.tsv'];
  mimeTypes = ['text/csv', 'text/tab-separated-values'];

  supportsFormat(filePath: string): boolean {
    const ext = filePath.toLowerCase().split('.').pop();
    return ext === 'csv' || ext === 'tsv';
  }

  load(filePath: string): ParserData[] {
    const validation = this.validateFile(filePath);
    if (!validation.isValid) {
      console.error(`Error: ${validation.error}`);
      process.exit(1);
    }

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const lines = content.trim().split('\n');
      
      if (lines.length === 0) {
        return [];
      }

      const delimiter = this.detectDelimiter(lines[0]);
      const headers = lines[0].split(delimiter);
      const data: ParserData[] = [];

      for (let i = 1; i < lines.length; i++) {
        const values = this.parseCSVLine(lines[i], delimiter);
        const row: ParserData = {};
        
        for (let j = 0; j < headers.length; j++) {
          const header = headers[j].trim();
          const value = values[j] !== undefined ? values[j].trim() : '';
          row[header] = this.inferType(value);
        }
        
        data.push(row);
      }

      return data;
    } catch (error) {
      console.error(`Error reading CSV: ${error instanceof Error ? error.message : String(error)}`);
      process.exit(1);
    }
  }

  private detectDelimiter(line: string): string {
    if (line.includes('\t')) return '\t';
    if (line.includes(',')) return ',';
    return ',';
  }

  private parseCSVLine(line: string, delimiter: string): string[] {
    const result: string[] = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
      const char = line[i];
      
      if (char === '"') {
        inQuotes = !inQuotes;
      } else if (char === delimiter && !inQuotes) {
        result.push(current);
        current = '';
      } else {
        current += char;
      }
    }
    
    result.push(current);
    return result;
  }

  private inferType(value: string): any {
    if (value === '') return null;
    
    if (!isNaN(Number(value)) && value.trim() !== '') {
      if (value.includes('.')) {
        return parseFloat(value);
      }
      return parseInt(value, 10);
    }
    
    return value;
  }
}
